### PR TITLE
Fix for redirect url, duplicated url in redirect. causes more 404's

### PR DESCRIPTION
**Pull request by tonypartridge**
Pull Request for Issue #17996 .

### Summary of Changes

Have an absolute 404 generated i.e.

http://localhost/artriclessssss.html
find in redirect manager and redirect to:
http://localhost/articles.html

404 redirected to:

http://localhost/http://localhost/articles.html

Expected result

Redirects to
http://localhost/articles.html

To fix, install patch. Test both absolutely and relative redirects work.

### CODE REVIEW

**Review by wilsonge — APPROVED** [review]
Looks good on review. Btw bugs introduced adds hours back onto the amount owed tony ;)

**Comment by ghost** [comment]
I have tested this item :white_check_mark: successfully on 3569fd6c9e4c523782bbf77ebcf25e335af29bf0<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17997">issues.joomla.org/tracker/joomla-cms/17997</a>.</sub>

**Review by HermanPeeren — COMMENTED** [review]
This fixes the symptom, not the cause why external URLS were wrongly redirected. The error is in line 282, where in strpos('http', $redirect->new_url) === false the needle and haystack is changed; it must be strpos($redirect->new_url, 'http') === false

**Comment by tonypartridge** [comment]
Great spot @HermanPeeren Updated now.

**Comment by ghost** [comment]
I have tested this item :white_check_mark: successfully on 915e25f3836d143e5bf2fd071f5e196b76f7ef9a<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17997">issues.joomla.org/tracker/joomla-cms/17997</a>.</sub>

**Comment by HermanPeeren** [comment]
If $redirect->new_url starts with 'index.php' then  JUri::isInternal($redirect->new_url) is true. Line 282 then sends the $redirect->new_url through JRoute::_(). No difference with what I'm suggesting, always going through JRoute::_().

The redirect-plugin supports relative urls only relative to the  JUri::root(). 

I have no example of a redirect that would go wrong if the $redirect->new_url goes trough JRoute::_(). Am I overlooking something?

**Comment by tonypartridge** [comment]
Exactly, so if it is True we want to parse it through JRoute as it is an internal URL, rather than redirect and then parsing if it's not already parsed.

**Comment by HermanPeeren** [comment]
Sorry, I don't understand your latest remark, @tonypartridge. 

- I suggested the conditional in line 282 might be superfluous and **we might just always go through JRoute**. 
- You said it was necessary for "support relative and non-relative urls so someone could start with index.php"
- I said that if a url starts with 'index.php' then JUri::isInternal() is true and so we go through JRoute; which is the same as I proposed (always going through JRoute).

**Do you have an example of a redirect-url that would go wrong if we always go through JRoute?**

**Comment by tonypartridge** [comment]
I haven't tested or checked it enough to know, but by doing a simple check there is nothing wrong with this code. It works and doesn't pass external urls through JRoute even if JRoute supports external URLs.

**Comment by HermanPeeren** [comment]
That is a very practical way to look at it. I agree with you that the current code does not give a wrong result.

So please ignore my esthetical code considerations. I'll try to find time to make unit tests and  refactor the code. For now the current code is useful and should be in J!3.8.1 as it repairs the 3.8.0 redirect mistake.

**Comment by tonypartridge** [comment]
Great thanks Herman! I agree :-)

-- 
Tony Partridge

On 21 September 2017 at 10:47:18, HermanPeeren (notifications@github.com)
wrote:

That is a very practical way to look at it. I agree with you that the
current code does not give a wrong result.

So please ignore my esthetical code considerations. I'll try to find time
to make unit tests and refactor the code. For now the current code is
useful and should be in J!3.8.1 as it repairs the 3.8.0 redirect mistake.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/joomla/joomla-cms/pull/17997#issuecomment-331107874>,
or mute the thread

File: plugins/system/redirect/redirect.php [hunk 1]
```diff
@@ -279,9 +279,8 @@ private static function doErrorHandling($error)
 					$redirect->new_url .= '?' . $urlQuery;
 				}
 
-				$dest = JUri::isInternal($redirect->new_url) || strpos('http', $redirect->new_url) === false ?
-					JRoute::_(JUri::root() . $redirect->new_url) :
-					$redirect->new_url;
+				$dest = JUri::isInternal($redirect->new_url) || strpos($redirect->new_url, 'http') === false ?
```
Comment (Fedik) [inline plugins/system/redirect/redirect.php:282]: need to test `http` only at 0 position, because it also can be at middle, as part of URL path (eg `/my-http-stuff`),
`strpos($redirect->new_url, 'http') !== 0` 
I think

Comment (HermanPeeren) [reply plugins/system/redirect/redirect.php:282]: @Fedik +1
(which gives an extra argument for leaving out code that would be superfluous, as it can introduce new errors, as demonstated here)

**Comment by tonypartridge** [comment]
So your argument is to skip everything and just chuck it all through
JRoute?

-- 
Tony Partridge

**Comment by HermanPeeren** [comment]
Yes. 
But be careful: my argument is solely based on reading the code and **results should be tested thoroughly**. That is why it is good to make tests before refactoring. Edge-cases like 'http' being part of the url but not on position 0 must be added to the testcases. As we have no unittests yet we should test the redirect-plugin for a variety of urls (internal, external, starting with 'index.php' or not etc.). Is it an idea to share such test-urls here?

**Comment by tonypartridge** [comment]
@mbabker when is 3.8.1 due out? Since this is a high priority issue, we
could merge as is then run tests and refactor for 3.8.2?

Or we do it all now

-- 
Tony Partridge

**Comment by HermanPeeren** [comment]
I would go for merging as is a.s.a.p. in 3.8.1, because a lot of sites have issues (like failing redirects) with 3.8.0. 

The error that @Fedik came with is also not very severe, because most likely those internal-urls with 'http' on a position >0 are likely to give JUri::isInternal()== true, so won't give much practical problems.

Then we have some more time to test and refactor with 3.8.2.

**Comment by tonypartridge** [comment]
I agree completely

-- 
Tony Partridge

**Comment by Ruud68** [comment]
I have tested this item :white_check_mark: successfully on 915e25f3836d143e5bf2fd071f5e196b76f7ef9a<br /><br />Hi, just found out this issue... It is killing my #SEO as redirects don't work anymore, also all my old URLs are not accessible anymore from facebook / twitter / and the likes :(
Tested this patch successfully! Thnx Tony!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17997">issues.joomla.org/tracker/joomla-cms/17997</a>.</sub>

**Comment by ghost** [comment]
RTC after two successful tests.

File: plugins/system/redirect/redirect.php (hunk 1 above)
Comment (gojets) [reply plugins/system/redirect/redirect.php:282]: Works for me. Thanks to all the contributors for working diligently on this issue!!!!
